src/buffer/out/textBuffer.cpp

```diff
@@ -2888,125 +2900,400 @@ std::vector<til::point_span> TextBuffer::SearchText(const std::wstring_view& nee
+    // first. Since StartPrompt actually sets up the prompt mark on the ROW, we
+    // need to do a bit of extra work here to start a new mark (if the last one
+    // wasn't in an appropriate state).
+
+    const auto mostRecentMarks = GetMarkExtents(1u);
+    if (!mostRecentMarks.empty())
     {
-        return;
+        const auto& mostRecentMark = til::at(mostRecentMarks, 0);
+        if (mostRecentMark.HasOutput())
+        {
+            // The most recent command mark had output. That suggests that either:
+            // * shell integration wasn't enabled (but the user would still
+            //   like lines with enters to be marked as prompts)
+            // * or we're in the middle of a command that's ongoing.
+
+            // If the mark doesn't have any command - then we know we're
+            // playing silly games with just marking whole lines as prompts,
+            // then immediately going to output.
+            //   --> add a new mark to this row, set all the attrs in this
+            //   row to be Prompt, and set the current attrs to Output.
+            //
+            // If it does have a command, then we're still in the output of
+            // that command.
+            //   --> the current attrs should already be set to Output.
+            if (!mostRecentMark.HasCommand())
+            {
+                auto& row = GetMutableRowByOffset(GetCursor().GetPosition().y);
+                row.StartPrompt();
+                return true;
+            }
+        }
+        else
+        {
+            // The most recent command mark _didn't_ have output yet. Great!
+            // we'll leave it alone, and just start treating text as Command or Output.
+        }
     }
-    auto& curr{ _marks.back() };
-    curr.end = pos;
-}
-void TextBuffer::SetCurrentCommandEnd(const til::point pos) noexcept
-{
-    if (_marks.empty())
+    else
     {
-        return;
+        // There were no marks at all!
+        //   --> add a new mark to this row, set all the attrs in this row
+        //   to be Prompt, and set the current attrs to Output.
+
+        auto& row = GetMutableRowByOffset(GetCursor().GetPosition().y);
+        row.StartPrompt();
+        return true;
     }
-    auto& curr{ _marks.back() };
-    curr.commandEnd = pos;
+    return false;
 }
-void TextBuffer::SetCurrentOutputEnd(const til::point pos, ::MarkCategory category) noexcept
+
+bool TextBuffer::StartCommand()
 {
-    if (_marks.empty())
+    const auto createdMark = _createPromptMarkIfNeeded();
+    auto attr = GetCurrentAttributes();
+    attr.SetMarkAttributes(MarkKind::Command);
+    SetCurrentAttributes(attr);
+    return createdMark;
+}
+bool TextBuffer::StartOutput()
+{
+    const auto createdMark = _createPromptMarkIfNeeded();
+    auto attr = GetCurrentAttributes();
+    attr.SetMarkAttributes(MarkKind::Output);
+    SetCurrentAttributes(attr);
+    return createdMark;
+}
+
+void TextBuffer::EndOutput(std::optional<unsigned int> error)

```

Comment:
comment may help here. we find the row above the cursor with the prompt and _then_ set the error code on it? can the function name do some more lifting here?